### PR TITLE
ENH: support non default indexes in writing to Parquet

### LINKED ISSUE #18581: Support non-default indexes in to_parquet
Calling `to_parquet` on a DataFrame with a non-default index results in the error below:
```python-traceback
ValueError: parquet does not support serializing a non-default index for the index; you can .reset_index() to make the index into column(s)
```
While, you can work-around this by calling `reset_index()`  as the message says, this loses the information about what columns made up the index so means you can't round-trip a DataFrame with a non-default index.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.21.0
pytest: 3.2.5
pip: 9.0.1
setuptools: 37.0.0
Cython: 0.27.3
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.7.1
xarray: None
IPython: 6.2.1
sphinx: 1.6.5
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: 1.4.4
```
</details>

### PR DESCRIPTION
**Pull request by dhirschfeld**
- [x] closes #18581
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
We need to check how this interacts with the `columns` keyword when reading back in. Ideally selecting only a few columns to read will still preserve your index (I think?), but I guess this is currently not happening (at least with pyarrow).

File: pandas/io/parquet.py
Comment (jschendel) [inline pandas/io/parquet.py]: Could create the `RangeIndex` via `RangeIndex(len(df))` here to make this a little cleaner.

File: pandas/io/parquet.py
Comment (jreback) [inline pandas/io/parquet.py]: so it *might* be better (though a bit more refactoring), to move the validation of index and validation of columns into methods on a base class to the impl

e.g.

```
class BaseImpl(object):


    def validate_index(....):
       ..

    def validate_columns(....):
      ....


class PyArrowImpl(BaseImpl):

     def write(...):
          if pyarrow < 0.7.1:
             self.validate_index()

         self.validate_columns()

class FastParquetImpl(BaseImpl):
      .....
```

as it makes things cleaner.

File: pandas/tests/io/test_parquet.py
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: could factor these 2 tests here from pyarrow impl (and also from fastparquet) into another test (that succeeds on all impls and versions)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some changes; pls. add a bug fix note in 0.22.0

File: pandas/io/parquet.py
Comment (dhirschfeld) [reply pandas/io/parquet.py]: Refactored in the latest commits...

File: pandas/tests/io/test_parquet.py
Comment (dhirschfeld) [reply pandas/tests/io/test_parquet.py]: Since the behaviour changes based on the library and the version of the library I'm not sure how else to do it while maintaining test coverage on all versions.

Feel free to push a fix if you'd like as I'm knocking off for the evening shortly...

File: pandas/io/parquet.py
Comment (jreback) [inline pandas/io/parquet.py]: import and use ``pandas.core.common.AbstractMethodError`` here instead

File: doc/source/whatsnew/v0.22.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: use ``:func:`DataFrame.to_parquet` `` here, use double-backticks around ``pyarrow>=0.7.0``

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments. if you can refactor the tests a bit ok (if not no big deal). ping on green.

File: pandas/io/parquet.py
Comment (dhirschfeld) [reply pandas/io/parquet.py]: Done.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
`fastparquet` also supports writing the index, so no need to still raise the errors in that case.

Also, to repeat my comment from before: we need to decide how this interacts with `columns` keyword before merging this.

**Comment by dhirschfeld** [comment]
> *`fastparquet` also supports writing the index, so no need to still raise the errors in that case.*

It seems it might just be easier all-round to just bump the pyarrow/fastparquet deps to the min required to support non-default indices - i.e. `>=0.7.0`  for pyarrow and ??? for fastparquet

**Comment by jreback** [comment]
would be ok with bumping pyarrow dep
i think fp is ok as is (i bumped it when doing this originally)

File: ci/requirements-3.5.sh
Comment (jreback) [inline ci/requirements-3.5.sh]: can you check if we have pyarrow installed in ci/* anywhere else (and if so update the pinned versions)

File: doc/source/whatsnew/v0.22.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: also there is another section on changed deps, pls add another entry which says pyarrow is now min 0.7.0 for parquet (also may need to update in io.rst)

**Review by jreback — CHANGES_REQUESTED** [review]
doc changes, ping on green.

**Comment by jorisvandenbossche** [comment]
@jreback not merge on green, see my request for changes on which I got no reply yet

BTW, I think we can do this for the 0.21.1 milestone. The other enhancements to `to/read_parquet` also went into that one (seeing it as fixes on very new functionality)

File: ci/requirements-3.5.sh
Comment (dhirschfeld) [reply ci/requirements-3.5.sh]: I checked and changed any pinned versions to the new minimum. If it wasn't already pinned I left it as is since a normal install will pull in later versions that the minimum.

Comment (dhirschfeld) [reply ci/requirements-3.5.sh]: image

**Comment by dhirschfeld** [comment]
@jorisvandenbossche - as of `fastparquet=0.1.3` it doesn't appear to support multi-indexes

image


...at least reading them anyway. Interestingly, it seems to write them fine and the file written by fastparquet can then be read in perfectly by pyarrow.

Not sure what to do about that but since fastparquet throws a sensible error I'd be inclined to just leave it and a future update will likely add the missing functionality.

**Comment by dhirschfeld** [comment]
@jorisvandenbossche - you're right about the `columns` argument not preserving the index:

image

The workaround, at the cost of efficiency, is simple enough:
```python
pd.io.parquet.read_parquet('tmp_pa.pq', engine='pyarrow')[['A','B']]
```
...but you have to know beforehand that the file was saved with a multi-index - i.e. it's a silent failure mode which isn't ideal.

If I name the index levels you can actually pull them out by name:
```python
df = pd.DataFrame(randn(2*len(dates), 3), columns=list('ABC'))
index = pd.MultiIndex.from_product([['Level1', 'Level2'], dates], names=['level','date'])
df.index = index
```
image
```python
pd.io.parquet.read_parquet('tmp_pa.pq', engine='pyarrow', columns=['A','B','level','date']).head()
```
image

...so, ideally if pandas can tell if the file has a multi-index the index names could then always be prepended to the columns argument.

**Comment by dhirschfeld** [comment]
@jorisvandenbossche - were there other changes you wanted that I've missed?

**Comment by jorisvandenbossche** [comment]
> Not sure what to do about that but since fastparquet throws a sensible error I'd be inclined to just leave it and a future update will likely add the missing functionality.

It's a bit unfortunate it only comes up when reading it back in, so either we are ok with that, or either we need to to raise on writing MIs with fastparquet.

> If I name the index levels you can actually pull them out by name:

That will not always work. Eg if they have no names, and pyarrow master has also changed this and the index levels will never have this name, so the workaround you show will not work there. 

The thing we can do is the following:

```
In [85]: import pyarrow.parquet as pq

In [86]: f = pq.ParquetFile("__test_mi.parquet")

In [87]: f.metadata.metadata[b'pandas']
Out[87]: b'{"columns": [{"name": "A", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "B", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "C", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "__index_level_0__", "pandas_type": "unicode", "numpy_type": "object", "metadata": null}, {"name": "__index_level_1__", "pandas_type": "datetime", "numpy_type": "datetime64[ns]", "metadata": null}], "index_columns": ["__index_level_0__", "__index_level_1__"], "pandas_version": "0.22.0.dev0+288.g6b6cfb8"}'

In [88]: import json

In [89]: pandas_metadata = json.loads(f.metadata.metadata[b'pandas'].decode('utf8'))

In [90]: pandas_metadata['index_columns']
Out[90]: ['__index_level_0__', '__index_level_1__']
```

and then add those to the `columns` list. 
And something similar for `fastparquet`:

```
In [98]: import fastparquet

In [99]: f = fastparquet.ParquetFile("__test_mi.parquet")

In [100]: json.loads(f.key_value_metadata['pandas'])['index_columns']
Out[100]: ['__index_level_0__', '__index_level_1__']
```

**Comment by dhirschfeld** [comment]
Heh, beat me to it:
```python
import pyarrow.parquet as pq
filemetadata = pq.read_metadata('tmp_fp.pq')
metadata = eval(filemetadata.metadata.get(b'pandas'), None, dict(null=None))
metadata['index_columns']
```
...eval is pretty nasty!

**Comment by jorisvandenbossche** [comment]
I also think that *if* there is a default index (range-like index without name, the case that was allowed now), we should not write it to the parquet file, because that would just be waste of time and space.

**Comment by jorisvandenbossche** [comment]
Yep, the `json.loads` is the better alternative :-)

**Comment by jreback** [comment]
> BTW, I think we can do this for the 0.21.1 milestone. The other enhancements to to/read_parquet also went into that one (seeing it as fixes on very new functionality)

since we are bumping the version I don't think this is a good idea. marked for 0.22

I am ok with this fix, @jorisvandenbossche

**Review by jreback — APPROVED** [review]

**Comment by dhirschfeld** [comment]
For pyarrow the columns fix is as simple as setting `use_pandas_metadata=True`:
```python
    def read(self, path, columns=None, **kwargs):
        path, _, _ = get_filepath_or_buffer(path)
        parquet_file = self.api.parquet.ParquetFile(path)
        kwargs['use_pandas_metadata'] = True
        return parquet_file.read(columns=columns, **kwargs).to_pandas()
```

For fastparquet I would use:
```python
    def read(self, path, columns=None, **kwargs):
        path, _, _ = get_filepath_or_buffer(path)
        parquet_file = self.api.ParquetFile(path)
        metadata = json.loads(parquet_file.key_value_metadata['pandas'])
        if columns is not None:
            columns = set(chain(columns, metadata['index_columns']))
        return parquet_file.to_pandas(columns=columns, **kwargs)
```
...but since it doesnt' support multi-indexes anyway it's a moot point.

**Comment by jreback** [comment]
@martindurant can you weigh in here on the fp fix?

**Comment by jorisvandenbossche** [comment]
> ...but since it doesnt' support multi-indexes anyway it's a moot point.

It's not a moot point, as it still supports single level index, which is better than nothing?

The fastparquet fix is fine, the MultiIndex 'not implemented' error is just an enhancement that would be welcome in fastparquet.

**Comment by jorisvandenbossche** [comment]
> I am ok with this fix, @jorisvandenbossche

I am not yet fine with the current version, so please don't merge.

> since we are bumping the version I don't think this is a good idea. marked for 0.22

If we do that, it's 0.21.1 that will become 0.22 and 0.22 will become 0.23 no?

**Comment by dhirschfeld** [comment]
I've pushed some changes to handle the columns issue - at least with pyarrow. The test is marked as xgail for fastparquet but that can be easily changed if fastparquet adds support for multi-indexes.

I think I've covered everything but let me know if there are further changes you'd like...

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: support -> supports

**Comment by jreback** [comment]
> if we do that, it's 0.21.1 that will become 0.22 and 0.22 will become 0.23 no?

huh?

no 0.21.1 - should release this soon as a bugfix, 0.22 for major changes.

File: pandas/io/parquet.py
Comment (dhirschfeld) [inline pandas/io/parquet.py]: For fastparquet I haven't bothered reading the metadata to pull out the columns since even if you do you still hit the `NotImplementedError`. I figure if fastparquet do add support for multi-indexes they might implement a solution similar to pyarrow so that the user doesn't have to introspect the metadata to reconstruct the correct DataFrame.

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: We should also test the index level names to be strings (if not None, and if written)

**Comment by jorisvandenbossche** [comment]
@dhirschfeld Thanks for the updates! Will review in a moment

**Comment by jorisvandenbossche** [comment]
> For fastparquet I haven't bothered reading the metadata to pull out the index columns 

Apparently it is doing this 'correctly' by default (`f` is a fastparquet ParquetFile):

```
In [142]: f.to_pandas()
Out[142]: 
    bar  baz  foo
zoo              
x     A    1  one
y     B    2  one
z     C    3  one
q     A    4  two
w     B    5  two
t     C    6  two

In [143]: f.to_pandas(columns=['bar'])
Out[143]: 
    bar
zoo    
x     A
y     B
z     C
q     A
w     B
t     C
```

So also there this "extracting metadata" is not needed

File: pandas/tests/io/test_parquet.py
Comment (jorisvandenbossche) [inline pandas/tests/io/test_parquet.py]: Is there a reason you can't use `check_round_trip` here?

File: pandas/tests/io/test_parquet.py
Comment (jorisvandenbossche) [inline pandas/tests/io/test_parquet.py]: Can you do the same test without names? (maybe parameterize)

**Comment by dhirschfeld** [comment]
Yep, fastparquet works fine with a non-multi-index DataFrame. The code I posted above would *also* work but was moot in the sense that it didn't fix the multi-index problem.

**Comment by jorisvandenbossche** [comment]
Ah, OK, then I understand your "moot" point :-)

**Review by jorisvandenbossche — COMMENTED** [review]
Looking good:

- Can you update the io.rst docs about parquet and index ?

**Comment by dhirschfeld** [comment]
Thanks for the reviews @jreback, @jorisvandenbossche & @jschendel!

Sorry it's taken a while to get here - I've usually only got a small window in the evenings when I can work on stuff not directly related to my *actual* work. That window is fast closing down this way so feel to push any changes in the interim. I'm also happy to pick it up again tomorrow though. No big rush if it's scheduled for 0.22...

**Comment by jorisvandenbossche** [comment]
Maybe not essential for this PR, but I think we should also think about:

- not writing the index if it is a default index (range-like index without name). It will be cheaper to just reconstruct it on reading (which is already done by both engines if there is no index column in the metadata)

- providing a `index=True/False` keyword similar to `to_csv` and others to control if you want to write the index or not (both pyarrow and fastparquet support this)

**Comment by jorisvandenbossche** [comment]
> Sorry it's taken a while to get here - I've usually only got a small window in the evenings when I can work on stuff not directly related to my actual work.

No problem! The PR is only open yet for 3 days :-) And it will be a very nice improvement.

**Comment by jorisvandenbossche** [comment]
@jreback to come back on the milestone. I personally would like to see this released as soon as possible, as it is an important change to make `read_parquet`/`to_parquet` as brand new functionality easier to use. This will impact how users use those functions, so let's change that as quick as possible. Since we are also including the addition of the `columns` keyword in the 0.21.1 release, I don't see why we wouldn't include this in the 0.21.1 release. 
(I can help with finishing this PR if time is a problem)

> since we are bumping the version I don't think this is a good idea. 

I don't fully understand what bumping the version has to do with it, but let's keep the discussion about version numbers for https://github.com/pandas-dev/pandas/issues/18244, and let's assume here that we want to release 0.21.1 in a couple of days (but as argued above, I would like to try to include it in that release).

**Comment by jreback** [comment]
well, almost all of the 0.21.1 fixes won't break your existing environment; that's the point.

This is actually something that will. that's why its 0.22.0. its an api breaking change. Now personally I agree that it would be fine to break this as its a new feature. If you want to be less strict on this because of that, then I am not opposed.

**Comment by jorisvandenbossche** [comment]
This will not really break code, it just enables a case that previously gave an error and people had to do workarounds (resetting the index), and thus will prevent that a lot of people that try this new functionality out have to do workarounds that will soon not be needed any more.

I know that I typically want to be rather strict on what gets included in a bug-fix release, but since this is brand new functionality + the other improvements to it already went in, I am fine with it in this case

**Comment by jreback** [comment]
no it *will* break code. if you have pyarrow < 0.7 installed, you are out of luck

**Comment by jreback** [comment]
> I know that I typically want to be rather strict on what gets included in a bug-fix release, but since this is brand new functionality + the other improvements to it already went in, I am fine with it in this case

what other improvements?

**Comment by jorisvandenbossche** [comment]
> no it will break code. if you have pyarrow < 0.7 installed, you are out of luck

ok, yes, for that part of the PR it is indeed breaking. But for the actual enhancement of enabling indexes (if you already had the latest pyarrow release) not.

> what other improvements?

The `columns` one and the "passing kwargs to engine" one.

**Comment by jorisvandenbossche** [comment]
> no it will break code. if you have pyarrow < 0.7 installed, you are out of luck

we could also split this PR -> add the ability to write index if pyarrow >= 0.7, and keep dropping support for arrow < 0.7 for a separate PR.

**Comment by dhirschfeld** [comment]
It would be possible to make it only fail in the read/write methods *if* there was a non-default index and pyarrow<0.7, rather than failing immediately in the constructor. 

Does that sound like a good compromise? If so I can take a crack at it today...

**Comment by jorisvandenbossche** [comment]
Yes, that would be a good compromise I think

**Comment by dhirschfeld** [comment]
It appears in `pyarrow=0.4.1` *reading* a multi-index DataFrame works
image
...though not columns, without the metadata introspection trick.

**Comment by dhirschfeld** [comment]
Had to squash the commits to get out of rebase hell!

I think I've addressed all comments and concerns so would be keen to get this in. As it is now it will still support pyarrow >= 0.4.1 up to the extent the library itself supports reading/writing parquet. 

TODO: Re-add the whatsnew entries. I wasn't sure if they should now go in v0.21.1 or not?

If there are other changes you'd like let me know...

**Comment by dhirschfeld** [comment]
I think the appveyor error is legit.
```
>       raise AssertionError(msg)
E       AssertionError: DataFrame.index are different
E       
E       Attribute "names" are different
E       [left]:  ['index']
E       [right]: [None]
```
On Windows at least fastparquet seems to name the index `index` even if there was no name for the index to start with.

**Comment by dhirschfeld** [comment]
The single Travis failure seems unrelated?

**Comment by jreback** [comment]
linting issues, you can see locally by: ``make lint-diff``

```
The command "ci/script_multi.sh" exited with 0.
124.03s$ ci/lint.sh
inside ci/lint.sh
Linting *.py
pandas/io/parquet.py:6:1: F401 'pandas.compat.range' imported but unused
pandas/io/parquet.py:90:12: E121 continuation line under-indented for hanging indent
pandas/io/parquet.py:117:5: E303 too many blank lines (2)
pandas/io/parquet.py:118:23: E128 continuation line under-indented for visual indent
pandas/io/parquet.py:123:80: E501 line too long (83 > 79 characters)
pandas/io/parquet.py:153:80: E501 line too long (83 > 79 characters)
pandas/tests/io/test_parquet.py:261:1: W293 blank line contains whitespace
pandas/tests/io/test_parquet.py:281:67: E241 multiple spaces after ','
pandas/tests/io/test_parquet.py:302:80: E501 line too long (81 > 79 characters)
pandas/tests/io/test_parquet.py:305:34: E226 missing whitespace around arithmetic operator
Linting *.py DONE
Linting setup.py
```

**Comment by dhirschfeld** [comment]
I added the whatsnew entry to 0.21.1 since the pyarrow version bump is no longer required.

If the version gets bumped in 0.22 then the crufty compatibility code can be removed then...

File: ci/requirements-2.7.sh
Comment (jorisvandenbossche) [inline ci/requirements-2.7.sh]: let's keep the old version here? (so we actually test the older version, now we dropped bumping the version in this PR)

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: Is there a reason you pass ` path, compression, coerce_timestamps` here?

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: Do we need to keep this for older pyarrow versions ?

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: maybe we should add here that they can also install latest pyarrow or fastparquet version (same for the others)

File: pandas/io/parquet.py
Comment (jorisvandenbossche) [inline pandas/io/parquet.py]: I would just pass `path` to the `_read_lt_070` function?

File: pandas/tests/io/test_parquet.py
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: we don't do the import like this anywhere else, appreciate changing it

File: pandas/tests/io/test_parquet.py
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: you can just pass pa instead of engine to skip fp

**Review by jorisvandenbossche — COMMENTED** [review]
@dhirschfeld Thanks for the update. I think there are still some left-overs to clean-up from the undo removing support for older versions, but I can also put some time in it this evening if needed.

File: pandas/io/parquet.py
Comment (dhirschfeld) [reply pandas/io/parquet.py]: Just a left over from a refactoring. Should be removed as they're no longer needed...

**Comment by jorisvandenbossche** [comment]
@dhirschfeld sorry for taking over your PR, but I want to get this merged for 0.21.1, and didn't know if you would have time today/tomorrow. 
I restored some of the testing code for testing the older versions, and fixed some bugs in the code to get those working. And addressed the comments of myself and @jreback.

But feel free to review my changes

**Comment by dhirschfeld** [comment]
@jorisvandenbossche - that's fine, I'd only be able to look into it tonight so am happy for you to take the lead!  I'm hoping to make use of it so am keen to get it in too...

**Comment by dhirschfeld** [comment]
LGTM!

**Comment by jorisvandenbossche** [comment]
And all CI is passing!

@jreback more comments? 
(the compat for older versions is maybe not the cleanest, but it will be directly removed on master anyway .. (I think we want to do that?))

File: pandas/io/parquet.py
Comment (jreback) [inline pandas/io/parquet.py]: can you add a blank line before comments, easier to read

**Review by jreback — APPROVED** [review]
tiny style comments, otherwise lgtm.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@dhirschfeld Thanks a lot for starting the PR!

**Comment by dhirschfeld** [comment]
Thanks @jorisvandenbossche for getting this over the line! I'm excited to make use of this new functionality! :)

**Pull request merged**
